runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/StageBundleFactory.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.fnexecution.control;
+
+/**
+ * A bundle factory scoped to a particular
+ * {@link org.apache.beam.runners.core.construction.graph.ExecutableStage}, which has all of the
+ * resources it needs to provide new {@link RemoteBundle RemoteBundles}.
+ *
+ * <p>Closing a StageBundleFactory signals that a a stage has completed and any resources bound to
+ * its lifetime can be cleaned up.
+ */
+public interface StageBundleFactory extends AutoCloseable {
+  /**
+   * Get a new {@link RemoteBundle bundle} for processing the data in an executable stage.
```

Comment:
in this stage

Reply:
fixed

Reply:
We were doing some weird munging in the hacking branch to correlate Targets with output destinations. This required reaching into `ProcessBundleDescriptor` in order to pull out the correct output tag: https://github.com/bsidhom/beam/blob/3e9b24a1247033e5d75f80e106d3d75052389799/runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java#L132

 This is no longer the correct abstraction to use now that we're concealing the bundle descriptor from the client. @tgroh  How else can we identify output nodes?

Reply:
Instead of going for a map of `target -> output receiver`, I'd go for a factory that consumes `ProcessBundleDescriptor` and `Target` which produces the appropriate `RemoteOutputReceiver`

Reply:
patched my commit to implement this.  Let me know if you had something different in mind.